### PR TITLE
[dagit] Tweak behavior of PartitionRangeInput

### LINKED ISSUE #3263: Partition selection via user-input strings has unexpected behavior
## Summary
I tried to launch a backfill for the first day of each month by inputting the string "2020-01-01, 2020-02-01, 2020-03-01, 2020-04-01, 2020-05-01, 2020-06-01, 2020-07-01, 2020-08-01, 2020-09-01, 2020-10-01, 2020-11-01" into the "Partitions" text box, copying the suggested format
Screen Shot 2020-11-21 at 1 38 53 PM

I expected the Preview section to be filled out and to be able to click "Launch" immediately, but neither happened. Seems like the "Preview" + "Launch" button doesn't refresh automatically; manually clicking into the Preview section causes the Preview section to populate + the "Launch" button is reactivated.

Could be nice to have a "force refresh / sync" button for users who manually enter strings. 

## Reproduction
Create a daily partition. Manually input the string "2020-01-01, 2020-02-01, 2020-03-01, 2020-04-01, 2020-05-01, 2020-06-01, 2020-07-01, 2020-08-01, 2020-09-01, 2020-10-01, 2020-11-01" into the "Partitions" text box

## Additional Info about Your Environment

https://www.loom.com/share/d45303d613424034a07868bd4b932ac4

### PR DESCRIPTION
**Pull request by hellendag**
## Summary

Resolves #3263.

Tweak behavior of `PartitionRangeInput`:

- Commit change on `Enter` as well as blur.
- Don't run the effect when `partitionNames` changes, since this is likely to happen on parent renders because it's a `string[]`. I noticed that the input sometimes blows away changes on re-renders, and we want to avoid this. Instead, track a stringified copy of the array to run the effect only when the `string[]` has actually given us something new.

## Test Plan

View "Launch Backfill" dialog. Add values to the input, commit them with `Enter`. Verify success. Add bogus value, try to commit, verify that the error appears. Repeat with `blur`, verify that this still works.

### CODE REVIEW

**Comment by hellendag** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6262**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

File: js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx [hunk 1]
```diff
@@ -9,29 +9,40 @@ export const PartitionRangeInput: React.FC<{
...
+    try {
+      onChange(textToPartitions(valueString, partitionNames));
+    } catch (err: any) {
+      e.preventDefault();
+      showCustomAlert({body: err.message});
+    }
+  };
+
+  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
```
Comment (bengotow) [inline js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx:33]: I think elsewhere in the app we use `Enter` or `Return` - want to do both here too? (I'm actually not sure which the return key is to be entirely honest, we may not need to do both anywhere.)

File: js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx [hunk 2]
```diff
@@ -9,29 +9,40 @@ export const PartitionRangeInput: React.FC<{
   partitionNames: string[];
 }> = ({value, onChange, partitionNames}) => {
   const [valueString, setValueString] = React.useState('');
+  const partitionNameJSON = React.useMemo(() => JSON.stringify(partitionNames), [partitionNames]);
 
   React.useEffect(() => {
-    setValueString(partitionsToText(value, partitionNames));
-  }, [value, partitionNames]);
+    const partitionNameArr = JSON.parse(partitionNameJSON);
```
Comment (bengotow) [inline js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx:15]: Ahh good catch I think this is a solid fix 👍  I think in the new backfills modal I built for assets the partition names are passed in already memoized, but it's probably not that way everywhere...

**Review by bengotow — APPROVED** [review]

File: js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx (hunk 1 above)
Comment (hellendag) [reply js_modules/dagit/packages/core/src/partitions/PartitionRangeInput.tsx:33]: I think we don't have to worry about `Return`, and `Enter` is sufficient.

**Pull request merged**